### DISCUSSION TITLE
Proposal: allow trailing comma for `params` arguments

### DISCUSSION THREAD

**Discussion by daniel-white** [dotnet/csharplang#1824] [Language Ideas]
Similar to #1246, but in a smaller scope.

Since `params` generates an array of the specified types from the remaining arguments, it is akin to a collection initializer. With no other arguments permitted after the `params` argument, this should be well defined as to which method is invoked. With logging or other format strings, this would be handy.

Given this signature:

```
 void Log(string format, params object[] args)
```

these should be treated the same

```
Log(
    format,
    a,
    b,
    c
);
```
and
```
Log(
    format,
    a,
    b,
    c,
);
```

**Comment by svick** [comment]
I don't think it makes sense to have special syntax just for `params`. And for the general case, there is already https://github.com/dotnet/csharplang/issues/391.

**Comment by jnm2** [comment]
https://github.com/dotnet/csharplang/issues/5354 would provide an alternative, using `[`...`]` which does support trailing commas instead of using `params`:

```cs
Log(format, [
    a,
    b,
    c,
]);
```

There's the existing, though slightly more visually distracting, array creation syntax:

```cs
Log(format, new[] {
    a,
    b,
    c,
});
```

**Reply by TonyValenti** [reply]
I am super excited about collection literals.  For me personally, once they get implemented, I'll likely refactor our libraries to remove ```params``` altogether.

**Reply by yaakov-h** [reply]
This solution won’t work once param Span comes along, and you might want to put some of your params back then

**Reply by jnm2** [reply]
@yaakov-h Why is that? We're trying to align span collection literals with span params.

**Reply by 333fred** [reply]
> This solution won’t work once param Span comes along, and you might want to put some of your params back then

I don't expect this to be true.

**Reply by yaakov-h** [reply]
oh did I completely forget about the target-typed part of literals at a quarter past midnight? whoopsie!
